File: src/main/java/tarmorn/playground/Functionality.java
package src.main.java.tarmorn.playground;

import java.util.ArrayList;
import java.util.HashSet;

import src.main.java.tarmorn.data.Triple;
import src.main.java.tarmorn.data.TripleSet;

public class Functionality {

	public static void main(String[] args) {
		
		 TripleSet ts = new TripleSet("data/FB15-237/train.txt");
		 
		 
		 HashSet<String> relations = new HashSet<String>();
		 relations.addAll(ts.getRelations());
		 
		 
		 
		 
		 for (String relation : relations) {
			System.out.println("relation: " + relation);
			ArrayList<Triple> rTriples = ts.getTriplesByRelation(relation); 
			HashSet<String> rHeads = new HashSet<String>();
			HashSet<String> rTails = new HashSet<String>();
			for (Triple rtriple : rTriples) {
				rHeads.add(rtriple.getHead());
				rTails.add(rtriple.getTail());
			}
			int tailsPerHeadAll = 0;
			int headsPerTailAll = 0;		
			for (String rh : rHeads) {
				int i = ts.getEntities(relation, rh, true).size();
				tailsPerHeadAll += i;
				
			}
			
			for (String rt : rTails) {
				int i = ts.getEntities(relation, rt, false).size();
				headsPerTailAll += i;
			}
			double headsPerTailFraction = (double)headsPerTailAll / (double)rTails.size();
			double tailsPerHeadFraction = (double)tailsPerHeadAll / (double)rHeads.size();
			System.out.println("   headsPerTailFraction: " + headsPerTailFraction);
			System.out.println("   tailsPerHeadFraction: " + tailsPerHeadFraction);
			
		 }
		
		

	}

}

File: src/main/java/tarmorn/playground/Functionality.kt
package tarmorn.playground

import tarmorn.data.Triple
import tarmorn.data.TripleSet

object Functionality {
    @JvmStatic
    fun main(args: Array<String>) {
        val ts = TripleSet("data/FB15-237/train.txt")


        val relations = HashSet<String>()
        relations.addAll(ts.relations)




        for (relation in relations) {
            println("relation: " + relation)
            val rTriples: MutableList<Triple> = ts.getTriplesByRelation(relation)
            val rHeads = HashSet<String>()
            val rTails = HashSet<String>()
            for (rtriple in rTriples) {
                rHeads.add(rtriple.h)
                rTails.add(rtriple.t)
            }
            var tailsPerHeadAll = 0
            var headsPerTailAll = 0
            for (rh in rHeads) {
                val i = ts.getEntities(relation, rh, true)!!.size
                tailsPerHeadAll += i
            }

            for (rt in rTails) {
                val i = ts.getEntities(relation, rt, false)!!.size
                headsPerTailAll += i
            }
            val headsPerTailFraction = headsPerTailAll.toDouble() / rTails.size.toDouble()
            val tailsPerHeadFraction = tailsPerHeadAll.toDouble() / rHeads.size.toDouble()
            println("   headsPerTailFraction: " + headsPerTailFraction)
            println("   tailsPerHeadFraction: " + tailsPerHeadFraction)
        }
    }
}
